common/include/pcl/common/impl/centroid.hpp

```diff
@@ -659,6 +660,270 @@ computeMeanAndCovarianceMatrix (const pcl::PointCloud<PointT> &cloud,
 }
 
 
+
+
+
+template <typename PointT, typename Scalar> inline unsigned int
+computeCentroidAndOBB (const pcl::PointCloud<PointT> &cloud,
+                  Eigen::Matrix<Scalar, 3, 1> &centroid,
+                  Eigen::Matrix<Scalar, 3, 1> &obb_position,
+                  Eigen::Matrix<Scalar, 3, 1> &obb_dimensions,
+                  Eigen::Matrix<Scalar, 3, 3> &obb_rotational_matrix)
+{
+    Eigen::Matrix<Scalar, 3, 3> covariance_matrix;
+    Eigen::Matrix<Scalar, 4, 1> centroid4;
+    unsigned int point_count= computeMeanAndCovarianceMatrix(cloud, covariance_matrix, centroid4);
+    if (!point_count)
+      return (0);
+    centroid(0) = centroid4(0);
+    centroid(1) = centroid4(1);
+    centroid(2) = centroid4(2);
+
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix<Scalar, 3, 3>> evd(covariance_matrix);
+
+    //Eigen::Vector<Scalar, 3> eigenvalues_ = evd.eigenvalues();
+    //major_value =    eigenvalues_(2); //promem order
+    //middle_value =   eigenvalues_(1);
+    //minor_value =    eigenvalues_(0);
+
+    Eigen::Matrix<Scalar, 3, 3> eigenvectors_ = evd.eigenvectors();
+    Eigen::Matrix<Scalar, 3, 1> major_axis;
+    Eigen::Matrix<Scalar, 3, 1> middle_axis;
+    Eigen::Matrix<Scalar, 3, 1> minor_axis;
+
+    minor_axis = eigenvectors_.col(0);//the eigenvectors do not need to be normalized (they are already)
+    middle_axis = eigenvectors_.col(1);
+
+    // cross product of the other two: ux = uy X uz   major = middle X minor
+    major_axis(0) = middle_axis(1) * minor_axis(2) - middle_axis(2) * minor_axis(1);
+    major_axis(1) = middle_axis(2) * minor_axis(0) - middle_axis(0) * minor_axis(2);
+    major_axis(2) = middle_axis(0) * minor_axis(1) - middle_axis(1) * minor_axis(0);
+
+    //when Scalar==double on a Windows 10 machine and MSVS:
+    //if you substitute the following Scalars with floats you get a 20% worse processing time, if with 2 PointT 55% worse
+    Scalar obb_min_pointx, obb_min_pointy, obb_min_pointz;
+    Scalar obb_max_pointx, obb_max_pointy, obb_max_pointz;
+    obb_min_pointx = obb_min_pointy = obb_min_pointz = std::numeric_limits<Scalar>::max();
+    obb_max_pointx = obb_max_pointy = obb_max_pointz = std::numeric_limits<Scalar>::min();
+
+    if (cloud.is_dense)
+    {
+      for (const auto& point : cloud)
+      {
+        Scalar xd = point.x - centroid[0], yd = point.y - centroid[1], zd = point.z - centroid[2];
+
+        Scalar x = xd * major_axis(0) + yd * major_axis(1) + zd * major_axis(2);
+        Scalar y = xd * middle_axis(0) + yd * middle_axis(1) + zd * middle_axis(2);
+        Scalar z = xd * minor_axis(0) + yd * minor_axis(1) + zd * minor_axis(2);
```

Comment:
I think this could be a `Eigen::Affine` transformation so that Eigen can easily use SIMD instructions.

Reply:
ok I'll check that